File: example-bots/cpp-sma-crossover/main_raw.cpp
/**
 * SMA Crossover Bot (C++)
 * =======================
 * A realtime bot that implements Simple Moving Average crossover strategy
 * using live data from Yahoo Finance.
 *
 * This example demonstrates:
 * - Fetching real market data from Yahoo Finance REST API using libcurl
 * - Calculating Simple Moving Averages (SMA)
 * - Detecting SMA crossovers for trading signals
 * - Structured metric emission for dashboard visualization
 *
 * Metrics emitted:
 * - price: Current stock price with change percentage
 * - sma: Short and long SMA values
 * - signal: BUY/SELL signals when crossover detected
 */

#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <chrono>
#include <thread>
#include <optional>
#include <cmath>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// Bot state
struct BotState {
    std::optional<double> prevShortSma;
    std::optional<double> prevLongSma;
};

// Callback for CURL to write response
static size_t WriteCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
    userp->append((char*)contents, size * nmemb);
    return size * nmemb;
}

// Round to specified decimal places
double roundTo(double value, int decimals) {
    double multiplier = std::pow(10.0, decimals);
    return std::round(value * multiplier) / multiplier;
}

// Get current timestamp as ISO string
std::string getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    auto epoch = now.time_since_epoch();
    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(epoch).count();
    auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(epoch).count() % 1000;
    return std::to_string(seconds) + "." + std::to_string(millis) + "Z";
}

// Emit a metric to stdout
void emitMetric(const std::string& metricType, json data) {
    data["_metric"] = metricType;
    std::cout << data.dump() << std::endl;
}

// Emit a log message to stdout
void emitLog(const std::string& event, json data) {
    data["event"] = event;
    data["timestamp"] = getCurrentTimestamp();
    std::cout << data.dump() << std::endl;
}

// Fetch data from Yahoo Finance
std::vector<double> fetchYahooFinance(CURL* curl, const std::string& symbol) {
    std::vector<double> prices;
    std::string url = "https://query1.finance.yahoo.com/v8/finance/chart/" + symbol + "?interval=1d&range=1mo";
    std::string response;

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "the0-sma-bot/1.0");
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);

    CURLcode res = curl_easy_perform(curl);
    if (res != CURLE_OK) {
        throw std::runtime_error(std::string("CURL error: ") + curl_easy_strerror(res));
    }

    // Parse JSON response
    json data = json::parse(response);

    if (data.contains("chart") &&
        data["chart"].contains("result") &&
        !data["chart"]["result"].empty()) {

        auto& result = data["chart"]["result"][0];
        if (result.contains("indicators") &&
            result["indicators"].contains("quote") &&
            !result["indicators"]["quote"].empty()) {

            auto& closeData = result["indicators"]["quote"][0]["close"];
            for (const auto& price : closeData) {
                if (!price.is_null()) {
                    prices.push_back(price.get<double>());
                }
            }
        }
    }

    return prices;
}

// Calculate Simple Moving Average
double calculateSMA(const std::vector<double>& prices, size_t period) {
    if (prices.size() < period) return 0.0;

    double sum = 0.0;
    for (size_t i = prices.size() - period; i < prices.size(); ++i) {
        sum += prices[i];
    }
    return sum / static_cast<double>(period);
}

// Check for crossover
std::optional<std::string> checkCrossover(double prevShort, double prevLong,
                                           double currShort, double currLong) {
    // Golden cross: short SMA crosses above long SMA
    if (prevShort <= prevLong && currShort > currLong) {
        return "BUY";
    }
    // Death cross: short SMA crosses below long SMA
    if (prevShort >= prevLong && currShort < currLong) {
        return "SELL";
    }
    return std::nullopt;
}

int main() {
    // Get configuration from environment
    std::string botId = std::getenv("BOT_ID") ? std::getenv("BOT_ID") : "test-bot";
    std::string configJson = std::getenv("BOT_CONFIG") ? std::getenv("BOT_CONFIG") : "{}";

    json config;
    try {
        config = json::parse(configJson);
    } catch (...) {
        config = json::object();
    }

    // Extract configuration with defaults
    std::string symbol = config.value("symbol", "AAPL");
    int shortPeriod = config.value("short_period", 5);
    int longPeriod = config.value("long_period", 20);
    int updateIntervalMs = config.value("update_interval_ms", 60000);

    emitLog("bot_started", {
        {"botId", botId},
        {"symbol", symbol},
        {"shortPeriod", shortPeriod},
        {"longPeriod", longPeriod}
    });

    // Initialize CURL
    curl_global_init(CURL_GLOBAL_DEFAULT);
    CURL* curl = curl_easy_init();
    if (!curl) {
        emitLog("error", {{"message", "Failed to initialize CURL"}});
        return 1;
    }

    BotState state;

    // Main loop
    while (true) {
        try {
            // Fetch historical data
            std::vector<double> prices = fetchYahooFinance(curl, symbol);

            if (prices.size() < static_cast<size_t>(longPeriod)) {
                emitLog("insufficient_data", {
                    {"symbol", symbol},
                    {"required", longPeriod},
                    {"available", prices.size()}
                });
                std::this_thread::sleep_for(std::chrono::milliseconds(updateIntervalMs));
                continue;
            }

            // Get current price
            double currentPrice = prices.back();
            double previousPrice = prices.size() > 1 ? prices[prices.size() - 2] : currentPrice;
            double changePct = previousPrice != 0.0 ?
                ((currentPrice - previousPrice) / previousPrice) * 100.0 : 0.0;

            // Emit price metric
            emitMetric("price", {
                {"symbol", symbol},
                {"value", roundTo(currentPrice, 2)},
                {"change_pct", roundTo(changePct, 3)},
                {"timestamp", getCurrentTimestamp()}
            });

            // Calculate SMAs
            double shortSma = calculateSMA(prices, shortPeriod);
            double longSma = calculateSMA(prices, longPeriod);

            // Emit SMA metric
            emitMetric("sma", {
                {"symbol", symbol},
                {"short_sma", roundTo(shortSma, 2)},
                {"long_sma", roundTo(longSma, 2)},
                {"short_period", shortPeriod},
                {"long_period", longPeriod}
            });

            // Check for crossover signal
            if (state.prevShortSma.has_value() && state.prevLongSma.has_value()) {
                auto signal = checkCrossover(
                    state.prevShortSma.value(),
                    state.prevLongSma.value(),
                    shortSma,
                    longSma
                );

                if (signal.has_value()) {
                    double confidence = std::min(std::abs(shortSma - longSma) / longSma * 100.0, 0.95);
                    std::string direction = signal.value() == "BUY" ? "above" : "below";

                    emitMetric("signal", {
                        {"type", signal.value()},
                        {"symbol", symbol},
                        {"price", roundTo(currentPrice, 2)},
                        {"confidence", roundTo(confidence, 2)},
                        {"reason", "SMA" + std::to_string(shortPeriod) + " crossed " + direction + " SMA" + std::to_string(longPeriod)}
                    });
                }
            }

            // Update previous SMA values
            state.prevShortSma = shortSma;
            state.prevLongSma = longSma;

        } catch (const std::exception& e) {
            emitLog("error", {{"message", e.what()}});
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(updateIntervalMs));
    }

    curl_easy_cleanup(curl);
    curl_global_cleanup();
    return 0;
}

File: example-bots/scala-sma-crossover/src/main/scala/Main_raw.scala
/**
 * SMA Crossover Bot (Scala 3)
 * ===========================
 * A realtime bot that implements Simple Moving Average crossover strategy
 * using live data from Yahoo Finance.
 *
 * This example demonstrates:
 * - Fetching real market data from Yahoo Finance REST API
 * - Calculating Simple Moving Averages (SMA) functionally
 * - Detecting SMA crossovers for trading signals
 * - Structured metric emission for dashboard visualization
 *
 * Metrics emitted:
 * - price: Current stock price with change percentage
 * - sma: Short and long SMA values
 * - signal: BUY/SELL signals when crossover detected
 */

import sttp.client3._
import io.circe._
import io.circe.parser._
import scala.util.{Try, Success, Failure}

case class BotState(
  prevShortSma: Option[Double] = None,
  prevLongSma: Option[Double] = None
)

object Main:
  def main(args: Array[String]): Unit =
    // Get configuration from environment
    val botId = sys.env.getOrElse("BOT_ID", "test-bot")
    val configJson = sys.env.getOrElse("BOT_CONFIG", "{}")

    val config = parse(configJson).getOrElse(Json.obj())

    // Extract configuration with defaults
    val symbol = config.hcursor.get[String]("symbol").getOrElse("AAPL")
    val shortPeriod = config.hcursor.get[Int]("short_period").getOrElse(5)
    val longPeriod = config.hcursor.get[Int]("long_period").getOrElse(20)
    val updateIntervalMs = config.hcursor.get[Int]("update_interval_ms").getOrElse(60000)

    emitLog("bot_started", Json.obj(
      "botId" -> Json.fromString(botId),
      "symbol" -> Json.fromString(symbol),
      "shortPeriod" -> Json.fromInt(shortPeriod),
      "longPeriod" -> Json.fromInt(longPeriod)
    ))

    val backend = HttpClientSyncBackend()
    var state = BotState()

    // Main loop
    while true do
      try
        state = processUpdate(backend, symbol, shortPeriod, longPeriod, state)
      catch
        case e: Exception =>
          emitLog("error", Json.obj("message" -> Json.fromString(e.getMessage)))

      Thread.sleep(updateIntervalMs)

  def processUpdate(
    backend: SttpBackend[Identity, Any],
    symbol: String,
    shortPeriod: Int,
    longPeriod: Int,
    state: BotState
  ): BotState =
    // Fetch historical data
    val prices = fetchYahooFinance(backend, symbol)

    if prices.length < longPeriod then
      emitLog("insufficient_data", Json.obj(
        "symbol" -> Json.fromString(symbol),
        "required" -> Json.fromInt(longPeriod),
        "available" -> Json.fromInt(prices.length)
      ))
      return state

    // Get current price
    val currentPrice = prices.last
    val previousPrice = if prices.length > 1 then prices(prices.length - 2) else currentPrice
    val changePct = if previousPrice != 0 then ((currentPrice - previousPrice) / previousPrice) * 100 else 0.0

    // Emit price metric
    emitMetric("price", Json.obj(
      "symbol" -> Json.fromString(symbol),
      "value" -> Json.fromDoubleOrNull(roundTo(currentPrice, 2)),
      "change_pct" -> Json.fromDoubleOrNull(roundTo(changePct, 3)),
      "timestamp" -> Json.fromString(getCurrentTimestamp)
    ))

    // Calculate SMAs
    val shortSma = calculateSMA(prices, shortPeriod)
    val longSma = calculateSMA(prices, longPeriod)

    // Emit SMA metric
    emitMetric("sma", Json.obj(
      "symbol" -> Json.fromString(symbol),
      "short_sma" -> Json.fromDoubleOrNull(roundTo(shortSma, 2)),
      "long_sma" -> Json.fromDoubleOrNull(roundTo(longSma, 2)),
      "short_period" -> Json.fromInt(shortPeriod),
      "long_period" -> Json.fromInt(longPeriod)
    ))

    // Check for crossover signal
    (state.prevShortSma, state.prevLongSma) match
      case (Some(prevShort), Some(prevLong)) =>
        checkCrossover(prevShort, prevLong, shortSma, longSma).foreach { signal =>
          val confidence = math.min(math.abs(shortSma - longSma) / longSma, 0.95)
          val direction = if signal == "BUY" then "above" else "below"

          emitMetric("signal", Json.obj(
            "type" -> Json.fromString(signal),
            "symbol" -> Json.fromString(symbol),
            "price" -> Json.fromDoubleOrNull(roundTo(currentPrice, 2)),
            "confidence" -> Json.fromDoubleOrNull(roundTo(confidence, 2)),
            "reason" -> Json.fromString(s"SMA$shortPeriod crossed $direction SMA$longPeriod")
          ))
        }
      case _ => ()

    // Update state
    state.copy(
      prevShortSma = Some(shortSma),
      prevLongSma = Some(longSma)
    )

  def fetchYahooFinance(backend: SttpBackend[Identity, Any], symbol: String): Vector[Double] =
    val url = s"https://query1.finance.yahoo.com/v8/finance/chart/$symbol?interval=1d&range=1mo"

    val request = basicRequest
      .get(uri"$url")
      .header("User-Agent", "the0-sma-bot/1.0")

    val response = request.send(backend)

    response.body match
      case Right(body) =>
        parse(body).toOption.flatMap { json =>
          val cursor = json.hcursor
          cursor
            .downField("chart")
            .downField("result")
            .downArray
            .downField("indicators")
            .downField("quote")
            .downArray
            .downField("close")
            .as[Vector[Option[Double]]]
            .toOption
            .map(_.flatten)
        }.getOrElse(Vector.empty)
      case Left(_) =>
        Vector.empty

  def calculateSMA(prices: Vector[Double], period: Int): Double =
    if prices.length < period then 0.0
    else prices.takeRight(period).sum / period

  def checkCrossover(
    prevShort: Double,
    prevLong: Double,
    currShort: Double,
    currLong: Double
  ): Option[String] =
    // Golden cross: short SMA crosses above long SMA
    if prevShort <= prevLong && currShort > currLong then Some("BUY")
    // Death cross: short SMA crosses below long SMA
    else if prevShort >= prevLong && currShort < currLong then Some("SELL")
    else None

  def emitMetric(metricType: String, data: Json): Unit =
    val output = data.deepMerge(Json.obj("_metric" -> Json.fromString(metricType)))
    println(output.noSpaces)

  def emitLog(event: String, data: Json): Unit =
    val output = data.deepMerge(Json.obj(
      "event" -> Json.fromString(event),
      "timestamp" -> Json.fromString(getCurrentTimestamp)
    ))
    println(output.noSpaces)

  def roundTo(value: Double, decimals: Int): Double =
    val multiplier = math.pow(10, decimals)
    math.round(value * multiplier) / multiplier

  def getCurrentTimestamp: String =
    val now = System.currentTimeMillis()
    s"${now / 1000}.${now % 1000}Z"
